src/analysis/normed_space/deriv.lean

```diff
@@ -17,40 +17,268 @@ is restricted to `s`. We also have
+* subtraction of two functions
+* bounded bilinear maps
+* multiplication of a function by a scalar function
+* multiplication of two scalar functions
+* composition of functions (the chain rule)
+
 -/
 import topology.basic topology.sequences topology.opens
-import analysis.normed_space.operator_norm
+import analysis.normed_space.operator_norm analysis.normed_space.bounded_linear_maps
 
 import analysis.asymptotics
 import tactic.abel
 
-open filter asymptotics bounded_linear_map
+open filter asymptotics bounded_linear_map set
+
+noncomputable theory
+local attribute [instance, priority 0] classical.decidable_inhabited classical.prop_decidable
 
 section
 
-variables {k : Type*} [normed_field k]
+set_option class.instance_max_depth 60
+
+variables {k : Type*} [nondiscrete_normed_field k]
 variables {E : Type*} [normed_space k E]
 variables {F : Type*} [normed_space k F]
 variables {G : Type*} [normed_space k G]
 
-variables (c : k)
-variables (f : E → F) (f' : E →L[k] F)
-variables (x : E) (s: set E) (L : filter E)
-
-def has_fderiv_at_filter :=
+def has_fderiv_at_filter (f : E → F) (f' : E →L[k] F) (x : E) (L : filter E) :=
 is_o (λ x', f x' - f x - f' (x' - x)) (λ x', x' - x) L
 
-def has_fderiv_at_within :=
+def has_fderiv_at_within (f : E → F) (f' : E →L[k] F) (x : E) (s : set E) :=
 has_fderiv_at_filter f f' x (nhds_within x s)
 
-def has_fderiv_at :=
+def has_fderiv_at (f : E → F) (f' : E →L[k] F) (x : E) :=
 has_fderiv_at_filter f f' x (nhds x)
 
-variables {f} {f₀ f₁ g: E → F}
-variables {f'} {f₀' f₁' g': E →L[k] F}
-variables {x}
-variables {s} {t : set E}
-variables {L} {L₁ L₂ : filter E}
+variables (k)
+
+def differentiable_at_within (f : E → F) (x : E) (s : set E) :=
+∃f' : E →L[k] F, has_fderiv_at_within f f' x s
+
+def differentiable_at (f : E → F) (x : E) :=
+∃f' : E →L[k] F, has_fderiv_at f f' x
+
+def fderiv_at_within (f : E → F) (x : E) (s : set E) : E →L[k] F :=
+if h : ∃f', has_fderiv_at_within f f' x s then classical.some h else 0
+
+def fderiv_at (f : E → F) (x : E) : E →L[k] F :=
+if h : ∃f', has_fderiv_at f f' x then classical.some h else 0
+
+def differentiable_on (f : E → F) (s : set E) :=
+  ∀x ∈ s, differentiable_at_within k f x s
+
+def differentiable (f : E → F) :=
+  ∀x, differentiable_at k f x
+
+/- A notation for sets on which the differential has to be unique. This is for instance the case
+on open sets, which is the main case of applications, but also
+on closed halfspaces or closed disks. It is only on such sets that it makes sense to talk about
+"the" derivative, and to talk about higher smoothness.
+
+The differential is unique when the tangent directions (called the tangent cone below) spans a
+dense subset of the underlying normed space. -/
+
+def tangent_cone_at (x : E) (s : set E) : set E :=
+  {y : E | ∃(c:ℕ → k), ∃(d: ℕ → E), {n:ℕ | x + d n ∈ s} ∈ (@at_top ℕ _) ∧
+    (tendsto (λn, ∥c n∥) at_top at_top) ∧ (tendsto (λn, c n • d n) at_top (nhds y))}
+
+def unique_diff_at_on (x : E) (s : set E) : Prop :=
```

Comment:
These definitions don't look at all like what their name suggests. I know you are going to prove that the name is justified, and you write this in a comment 15 lines lower. But wouldn't it make sense to leave the name closer to the actual `def`? Maybe something like `span_tangent_cone_at_dense`?
Otherwise please also add a short comment here pointing out that you prove that the name is justified.